include/data_struct.h

```diff
@@ -134,15 +134,29 @@ class BaseDeltaStates {
     virtual void swap(BaseDeltaStates &other);
 };
 
+class BaseTempSLinear {
+   public:
+    std::vector<float> cov_hh, mu_h_prev;
+    std::vector<float> mu_zo_smooths, var_zo_smooths;
+    size_t num_hs = 0;
+
+    BaseTempSLinear(size_t num_hs);
+    BaseTempSLinear();
+    ~BaseTempSLinear() = default;
+    // virtual void set_size(size_t size, size_t size_w);
+    virtual void set_num_states(size_t num_hs);
+    void reset_zeros();
+};
+
 class BaseTempStates {
    public:
     std::vector<float> tmp_1;
     std::vector<float> tmp_2;
     size_t size = 0, block_size = 0, actual_size = 0;
+    BaseTempSLinear slinear;
```

Comment:
This is not a good idea...This class is a special case for smoother. It should be embedded in SLinear or SLSTM

Reply:
Given a network (SLSTM, SLinear), in the forward pass, I need to transfer some variables from SLSTM to SLinear. I see two possible ways to do it: (1) modify `BaseTempStates` as I have done. (2) Modify the `BaseHiddenStates` that affects the `input_z_buffer` in `Sequential` class. What do you suggest?

Reply:
None of them look good to me! The code will become bloated. `Base` means that the variables of that class will be  used more than >50% and other data class can be derived for it. 
Here is an example I've done for [cuda version](https://github.com/lhnguyen102/cuTAGI/blob/main/include/data_struct_cuda.cuh#L18). For such case, a dispatch mechanism is required [here](https://github.com/lhnguyen102/cuTAGI/blob/main/src/sequential.cpp#L121).  

In this case, I think the ideal case is to define another data class `SmoothingHiddenStates` that are derived from `BaseHiddenStates` because it will be able to pass through `.forward(...)` function of each layer. Within that layer, you can use `dynamic_cast` to have access to your smoothing variables of the class. Before doing that, you will need to ask the following questions:
1. Which variables will be used for `backward(...)` and `smoother(...)`? These variables will be stored in `SLinear` or `SLSTM` classes because they will be called during backward pass and smoother after the forward pass is done See example [here](https://github.com/lhnguyen102/cuTAGI/blob/main/include/data_struct.h#L150) 
2. Which variables will be used to pass the info/data between layers? these have to be stored in a class derived from `BaseHiddenStates` in order to go through the `forward(...)`.  Buffer and temp states are supposed to hold the data between layer_i and layer_i+1. If it goes beyond that, it won't work.

Reply:
For the case of cuda version `HiddenStateCuda` is derived from `BaseHiddenStates`
```c++
class HiddenStateCuda : public BaseHiddenStates
```
In this case, I will need all the variables of `BaseHiddenStates` and additional variables 
```c++
    float *d_mu_a = nullptr;
    float *d_var_a = nullptr;
    float *d_jcb = nullptr
```

Reply:
Is there any potential use of model like this
```c++
Sequential(
  Linear(...),
  SLSTM(...),
  SLinear(...),
  Linear(...),
)
```
If so, my proposal wont work

Reply:
If this is the case, you have no choice but adding those smoothing variables directly to `BaseHiddenStates`. The latter will lead to unused variables and waste memory resources for all the cases that does not need the smoothing pass

Reply:
> Is there any potential use of model like this
> 
> ```c++
> Sequential(
>   Linear(...),
>   SLSTM(...),
>   SLinear(...),
>   Linear(...),
> )
> ```
> 
> If so, my proposal wont work

No. We won't use this.

Reply:
I will create a new data class `SmoothingHiddenStates` that is derived from `BaseHiddenStates` as you suggest. Thanks.